### PR TITLE
Improvement/cldsrv 168 delete bucket tagging

### PR DESCRIPTION
**Pull request by ghost**
Monitoring not supported on 7.x branches.

### CODE REVIEW

**Review by ghost — APPROVED** [review]
LGTM

File: lib/api/bucketDeleteTagging.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        (bucket, next) => {
+            bucket.setTags([]);
+            metadata.updateBucket(bucket.getName(), bucket, log, err =>
+                next(err, bucket));
+        },
+    ], (err, bucket) => {
+        const corsHeaders = collectCorsHeaders(request.headers.origin,
+            request.method, bucket);
+        return callback(err, corsHeaders);
+    });
```
Comment (ghost) [inline lib/api/bucketDeleteTagging.js]:
```suggestion
    ], (err, bucket) => {
        const corsHeaders = collectCorsHeaders(request.headers.origin,
            request.method, bucket);
        pushMetric('deleteBucketTagging', log, {
            authInfo,
            bucket: bucketName,
        });
        return callback(err, corsHeaders);
    });
```

**Review by ghost — COMMENTED** [review]
Suggested a missing pushMetrics call

**Review by KillianG — CHANGES_REQUESTED** [review]
I think you are missing tests for getBucketTagging no ?

**Comment by ghost** [comment]
@KillianG they are in the GET PR AFAIK, as Get is needed in the tests.

**Comment by KillianG** [comment]
> @KillianG they are in the GET PR AFAIK, as Get is needed in the tests.

Oops I meant get bucket tagging, edited

**Review by nicolas2bert — COMMENTED** [review]
Why is there no test?

**Comment by ghost** [comment]
@nicolas2bert The tests are added in [getBucketTagging](https://github.com/scality/cloudserver/pull/4481) as they would be incomplete otherwise

File: lib/api/bucketDeleteTagging.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    return waterfall([
+        next => metadataValidateBucket(metadataValParams, log,
+            (err, bucket) => next(err, bucket)),
+        (bucket, next) => {
+            bucket.setTags([]);
+            metadata.updateBucket(bucket.getName(), bucket, log, err =>
+                next(err, bucket));
+        },
+    ], (err, bucket) => {
+        const corsHeaders = collectCorsHeaders(request.headers.origin,
```
Comment (nicolas2bert) [inline lib/api/bucketDeleteTagging.js]: log the error in case of failure?

File: lib/api/bucketDeleteTagging.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            request.method, bucket);
+        pushMetric('deleteBucketTagging', log, {
```
Comment (nicolas2bert) [inline lib/api/bucketDeleteTagging.js:36]: Should we push UTAPI metrics even if the call failed?

**Review by KillianG — APPROVED** [review]

File: lib/api/bucketDeleteTagging.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Bucket Delete Tagging - Delete a bucket's Tagging
+ * @param {AuthInfo} authInfo - Instance of AuthInfo class with requester's info
+ * @param {object} request - http request object
+ * @param {object} log - Werelogs logger
+ * @param {function} callback - callback to server
+ * @return {undefined}
+ */
+function bucketDeleteTagging(authInfo, request, log, callback) {
```
Comment (nicolas2bert) [inline lib/api/bucketDeleteTagging.js]: Is there a plan for monitoring this call?

File: lib/api/bucketDeleteTagging.js (hunk 2 above)
Comment (ghost) [reply lib/api/bucketDeleteTagging.js]: Is this ok? https://github.com/scality/cloudserver/pull/4480/commits/94700611b434aa8bb2dc5c21bf7071f44d328a89#diff-ecb34075d5365a2b604ea4bcb794846a9c3352a7bdbcd5caebbb8a2c7862466eR37

File: lib/api/bucketDeleteTagging.js (hunk 3 above)
Comment (ghost) [reply lib/api/bucketDeleteTagging.js:36]: moved out of error scope, this ok? https://github.com/scality/cloudserver/pull/4480/commits/94700611b434aa8bb2dc5c21bf7071f44d328a89#diff-ecb34075d5365a2b604ea4bcb794846a9c3352a7bdbcd5caebbb8a2c7862466eR41

File: lib/api/bucketDeleteTagging.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            bucket.setTags([]);
+            metadata.updateBucket(bucket.getName(), bucket, log, err =>
+                next(err, bucket));
+        },
+    ], (err, bucket) => {
+        const corsHeaders = collectCorsHeaders(request.headers.origin,
+            request.method, bucket);
+        if (err) {
+            log.trace('error processing request', { error: err,
+                method: 'deleteBucketTagging' });
```
Comment (ghost) [inline lib/api/bucketDeleteTagging.js]: I feel like we should use a `log.error` here, as the error can be a metadata error. What do you think?

Comment (ghost) [reply lib/api/bucketDeleteTagging.js]: https://github.com/scality/cloudserver/pull/4480/commits/8e1820362ca258dae30885496ef79cb79282f0d0#diff-ecb34075d5365a2b604ea4bcb794846a9c3352a7bdbcd5caebbb8a2c7862466eR37

File: lib/api/bucketDeleteTagging.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    const bucketName = request.bucketName;
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        requestType: 'bucketDeleteTagging',
+    };
+
+    try {
+        let promisifyFunc = util.promisify(metadataValidateBucket);
+        bucket = await promisifyFunc(metadataValParams, log);
```
Comment (ghost) [inline lib/api/bucketDeleteTagging.js]: Maybe you already discussed that with @miniscruff but I am not a big fan of try/catch to catch Promise errors...
Why not trying to stick with golang-like logic using https://www.npmjs.com/package/await-to-js ?

Comment (ghost) [reply lib/api/bucketDeleteTagging.js]: Otherwise LGTM

Comment (miniscruff) [reply lib/api/bucketDeleteTagging.js]: Hmm, not sure the issue with try/catch but for JS it is definitely the standard and most common approach. I think we might want to make our try catches smaller, as in one around each thing that can error. But I would not be a fan of using that package, don't get me wrong, I really like go syntax but for JS it will put us on an island as it is not standard at all.

File: lib/api/bucketDeleteTagging.js (hunk 3 above)
Comment (miniscruff) [reply lib/api/bucketDeleteTagging.js:36]: Resolving Nicolas's comments as he is on a long vacation.

File: package.json [hunk 7]
```diff
@@ -42,7 +42,7 @@
   },
   "devDependencies": {
     "bluebird": "^3.3.1",
-    "eslint": "^2.4.0",
+    "eslint": "^3.6.0",
```
Comment (miniscruff) [inline package.json]: not sure why this change was added to this PR, adds a bit of fluff

File: lib/api/bucketDeleteTagging.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    let bucket;
+    const bucketName = request.bucketName;
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        requestType: 'bucketDeleteTagging',
+    };
+
+    try {
+        let promisifyFunc = util.promisify(metadataValidateBucket);
```
Comment (miniscruff) [inline lib/api/bucketDeleteTagging.js]: Same comment from the other PR

**Review by miniscruff — CHANGES_REQUESTED** [review]

File: package.json (hunk 7 above)
Comment (ghost) [reply package.json]: because otherwise [this line ](https://github.com/scality/cloudserver/pull/4480/files#diff-ecb34075d5365a2b604ea4bcb794846a9c3352a7bdbcd5caebbb8a2c7862466eR33) triggers the linter. Do you suggest putting an ignore comment instead?

File: lib/api/bucketDeleteTagging.js (hunk 8 above)
Comment (ghost) [reply lib/api/bucketDeleteTagging.js]: done https://github.com/scality/cloudserver/pull/4480/files#diff-ecb34075d5365a2b604ea4bcb794846a9c3352a7bdbcd5caebbb8a2c7862466eR19

**Review by KillianG — APPROVED** [review]

File: package.json (hunk 7 above)
Comment (miniscruff) [reply package.json]: What is wrong with that line? I think an ignore is easier if it can't be fixed with a follow up later. Generally upgrading eslint gives a lot of changes, this one was sort of small but still a bit.

File: lib/api/bucketDeleteTagging.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        bucket = await metadataValidateBucketPromise(metadataValParams, log);
+        bucket.setTags([]);
+        await updateBucketPromise(bucket.getName(), bucket, log);
+        const corsHeaders = collectCorsHeaders(request.headers.origin,
+            request.method, bucket);
+        pushMetric('deleteBucketTagging', log, {
+            authInfo,
+            bucket: bucketName,
+        });
+        return callback(null, corsHeaders);
```
Comment (miniscruff) [inline lib/api/bucketDeleteTagging.js]: Not really asking for change here as we do it all the time but the return is not required. We use it as an early return and in this case it won't matter. Although, if our callback throws an error we might do something we don't want to.
Generally I prefer to do
```js
callback(...);
return;
```
But its fine.

**Review by miniscruff — APPROVED** [review]

File: package.json (hunk 7 above)
Comment (jonathan-gramain) [reply package.json]: I agree with @miniscruff, in this PR's scope it's better not to upgrade the linter version and just add an ignore rule if the line is legit/needed and cannot easily be changed to please the linter.

If we decide that we should upgrade the linter version, then we can do it with a separate ticket and PR.

File: lib/api/bucketDeleteTagging.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Bucket Delete Tagging - Delete a bucket's Tagging
+ * @param {AuthInfo} authInfo - Instance of AuthInfo class with requester's info
+ * @param {object} request - http request object
+ * @param {object} log - Werelogs logger
+ * @param {function} callback - callback to server
+ * @return {undefined}
+ */
+async function bucketDeleteTagging(authInfo, request, log, callback) {
+    log.debug('processing request', { method: 'bucketDeleteTagging' });
```
Comment (jonathan-gramain) [inline lib/api/bucketDeleteTagging.js]: This log may be more useful if also logging some request info, at least the bucket name and maybe also the owner ID

File: lib/api/bucketDeleteTagging.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+            request.method, bucket);
+        pushMetric('deleteBucketTagging', log, {
+            authInfo,
+            bucket: bucketName,
+        });
+        return callback(null, corsHeaders);
+    } catch (err) {
+        const corsHeaders = collectCorsHeaders(request.headers.origin,
+            request.method, bucket);
+        log.error('error processing request', { error: err,
```
Comment (jonathan-gramain) [inline lib/api/bucketDeleteTagging.js:42]: same here with logging the bucket name

File: lib/api/bucketDeleteTagging.js (hunk 9 above)
Comment (jonathan-gramain) [inline lib/api/bucketDeleteTagging.js]: I'm not sure it is a real issue here, but at least something to check: we might want to be wary of exceptions that may be raised from within the callback, in order to avoid them being caught and the callback being called again with an error. In doubt, I suggest to move the callback after the `try...catch` block, and `return` in the `catch` block.

In the callback-based code base, we usually do not throw exceptions on purpose, and if they occur anyway, we do not catch them so they end up crashing the worker. If we make use of exceptions and attempt to catch them, this kind of issue of double callback becomes a possibility.

Comment (jonathan-gramain) [inline lib/api/bucketDeleteTagging.js:40]: I'm not sure it is a real issue here, but at least something to check: we might want to be wary of exceptions that may be raised from within the callback, in order to avoid them being caught and the callback being called again with an error. In doubt, I suggest to move the callback after the `try...catch` block, and `return` in the `catch` block.

In the callback-based code base, we usually do not throw exceptions on purpose, and if they occur anyway, we do not catch them so they end up crashing the process. If we make use of exceptions and attempt to catch them, this kind of issue of double callback becomes a possibility.

Comment (miniscruff) [reply lib/api/bucketDeleteTagging.js]: oh I mentioned something similar in one of my comments, so at least I am on the same page sometimes. :smirk:

**Review by watsaqat — APPROVED** [review]

**Review by watsaqat — APPROVED** [review]

File: lib/api/bucketDeleteTagging.js (hunk 9 above)
Comment (ghost) [reply lib/api/bucketDeleteTagging.js]: done 684e92cd5c1651787050c6a7bb8e8f440cb8ce2f

File: lib/api/bucketDeleteTagging.js (hunk 10 above)
Comment (ghost) [reply lib/api/bucketDeleteTagging.js]: done 684e92cd5c1651787050c6a7bb8e8f440cb8ce2f

File: lib/api/bucketDeleteTagging.js (hunk 11 above)
Comment (ghost) [reply lib/api/bucketDeleteTagging.js:42]: done 684e92cd5c1651787050c6a7bb8e8f440cb8ce2f

**Comment by ghost** [comment]
ping
